# Multicollinearity in Linear Regression

Correlation is a statistical measure that expresses the extent to which two variables are linearly related (meaning they change together at a constant rate).
For example, being educated might negatively correlate with the crime rate when an increase in one variable leads to a decrease in another and vice versa. If a country's education level is improved, it can lower crime rates

Likewise in multicollinearity there consist of multiple features which effect.

In [12]:
import pandas as pd
import numpy as np
import statsmodels.api as sm
df = pd.read_csv("Advertising.csv",index_col=0)
df.head()

TV  radio  newspaper  sales
1  230.1   37.8       69.2   22.1
2   44.5   39.3       45.1   10.4
3   17.2   45.9       69.3    9.3
4  151.5   41.3       58.5   18.5
5  180.8   10.8       58.4   12.9

In [13]:
x = df[['TV','radio','newspaper']]
y=df[['sales']]

in linear regression : y = Bo + B1(TV) + B2(radio)+B3(newspaper)
but Bo is not included here. so adding it on the dataset which is constant all over the dataset 1

In [14]:
### fit a OLS(ordinary linear squared) model with intercept on TV and radio
x = sm.add_constant(x)
model = sm.OLS(y,x).fit()

In [15]:
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  sales   R-squared:                       0.897
Model:                            OLS   Adj. R-squared:                  0.896
Method:                 Least Squares   F-statistic:                     570.3
Date:                Sat, 20 Jan 2024   Prob (F-statistic):           1.58e-96
Time:                        19:43:31   Log-Likelihood:                -386.18
No. Observations:                 200   AIC:                             780.4
Df Residuals:                     196   BIC:                             793.6
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.9389      0.312      9.422      0.000       2.324       3.554
TV             0.0458      0.001     32.809      0.000       0.043       0.049
radio          0.1885      0.009     21.893      0.000       0.172       0.206
newspaper     -0.0010      0.006     -0.177      0.860      -0.013       0.011
==============================================================================
Omnibus:                       60.414   Durbin-Watson:                   2.084
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              151.241
Skew:                          -1.327   Prob(JB):                     1.44e-33
Kurtosis:                       6.332   Cond. No.                         454.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

First check the coefficent of independent features, it showing very less number and then R2 which show good (the number near to 1 means the model is good) and then check standard error
there is no high number so the multicollinearity doesn't exist here

In [16]:
import matplotlib.pyplot as plt
x.iloc[:,1:].corr()

TV     radio  newspaper
TV         1.000000  0.054809   0.056648
radio      0.054809  1.000000   0.354104
newspaper  0.056648  0.354104   1.000000

Using this method we can check the collinearity,  the reationship between the independent featues is pretty much nearly equal to zero. So the Multicollinearity doesn't exist

In [17]:
df_salary = pd.read_csv('Salary_data.csv')
df_salary.head()

YearsExperience   Age  Salary
0              1.1  21.0   39343
1              1.3  21.5   46205
2              1.5  21.7   37731
3              2.0  22.0   43525
4              2.2  22.2   39891

In [18]:
x = df_salary[['YearsExperience','Age']]
y = df_salary[['Salary']]

In [19]:
# fit a ols model with intercept on YearsExperince and age


In [20]:
x = sm.add_constant(x)
model = sm.OLS(y,x).fit()

In [21]:
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 Salary   R-squared:                       0.960
Model:                            OLS   Adj. R-squared:                  0.957
Method:                 Least Squares   F-statistic:                     323.9
Date:                Sat, 20 Jan 2024   Prob (F-statistic):           1.35e-19
Time:                        19:43:40   Log-Likelihood:                -300.35
No. Observations:                  30   AIC:                             606.7
Df Residuals:                      27   BIC:                             610.9
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
===================================================================================
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
const           -6661.9872   2.28e+04     -0.292      0.773   -5.35e+04    4.02e+04
YearsExperience  6153.3533   2337.092      2.633      0.014    1358.037    1.09e+04
Age              1836.0136   1285.034      1.429      0.165    -800.659    4472.686
==============================================================================
Omnibus:                        2.695   Durbin-Watson:                   1.711
Prob(Omnibus):                  0.260   Jarque-Bera (JB):                1.975
Skew:                           0.456   Prob(JB):                        0.372
Kurtosis:                       2.135   Cond. No.                         626.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Now in model summary when we check the standard error it'showing the high value and after checking the p value of yearsexp is less than 0.05(considering the 95% CF)but for age is greater
so we can assume there may be correlation exist To conform let's plot

In [22]:
x.iloc[:,1:].corr()

YearsExperience       Age
YearsExperience         1.000000  0.987258
Age                     0.987258  1.000000

Here  The correlation between yearsExpereince and age is 98% which is huge. 
for solving the problem we can remove the one of the features to predict the salary.
